databroker/core.py

```diff
@@ -1273,13 +1288,29 @@ def __init__(self,
 
         self._run_stop_doc = metadata['stop']
         self._run_start_doc = metadata['start']
-        self._descriptors = [descriptor for descriptor in metadata['descriptors']
-                             if descriptor.get('name') == self._stream_name]
-        # Should figure out a way so that self._resources doesn't have to be
-        # all of the Run's resources.
-        self._resources = metadata['resources']
         self._partitions = None
 
+    def _load(self):
+        # TODO Add driver API to fetch only the descriptors of interest instead
+        # of fetching all of them and then filtering.
+        self._descriptors = [Descriptor(self._transforms['descriptor'](descriptor))
+                             for descriptor in self._get_event_descriptors()
+                             if descriptor.get('name') == self._stream_name]
+        # TODO Should figure out a way so that self._resources doesn't have to
+        # be all of the Run's resources.
+        self._resources = [Resource(self._transforms['resource'](resource))
+                           for resource in self._get_resources()]
+        
+        # get_run_stop() may return None if the document was never created due
+        # to a critical failure or simply not yet emitted during a Run that is
+        # still in progress. If it returns None, pass that through.
+        stop = self._get_run_stop()
```

Comment:
Can we first check if `self._run_stop_doc is None`? Unlike descriptors and resources, once we have a stop document, we don't need to check back to see if any more are coming.

Reply:
I copied this from the Run, should I change it in both places?

Reply:
Yes, thanks.

Reply:
updated